File: docs/queue/code/queue.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

typedef struct {
    Node *front;
    Node *back;
    int size;
} Queue;

Queue create_queue() {
    Queue q;
    q.front = NULL;
    q.back = NULL;
    q.size = 0;
    return q;
}

void enqueue(Queue *q, int data) {
    Node *node = (Node *)malloc(sizeof(Node));
    node->data = data;
    node->next = NULL;

    if (q->back == NULL) {
        q->front = node;
        q->back = node;
    } else {
        q->back->next = node;
        q->back = node;
    }

    q->size++;
}

int dequeue(Queue *q) {
    if (q->front == NULL) {
        printf("Queue is empty\n");
        return -1;
    }

    Node *temp = q->front;
    int data = temp->data;

    q->front = q->front->next;

    if (q->front == NULL) {
        q->back = NULL;
    }

    free(temp);
    q->size--;

    return data;
}

int main() {
    Queue q = create_queue();
    enqueue(&q, 10);     // FRONT -> [10] <- BACK
    enqueue(&q, 20);     // FRONT -> [10] -> [20] <- BACK
    enqueue(&q, 30);     // FRONT -> [10] -> [20] -> [30] <- BACK

    int a = dequeue(&q);    // a = 1, FRONT -> [20] -> [30] <- BACK
    int b = dequeue(&q);    // b = 2, FRONT -> [30] <- BACK
    int c = dequeue(&q);    // c = 3, FRONT -> NULL <- BACK

    printf("Dequeued values: %d, %d, %d\n", a, b, c); // Output: Dequeued values: 10, 20, 30
    return 0;
}

File: docs/queue/code/queue.rs
struct Node<T> {
    data: T,
    next: Option<Box<Node<T>>>,
}

struct Queue<T> {
    front: Option<Box<Node<T>>>,
    back: *mut Node<T>,
    size: usize,
}

impl<T> Queue<T> {
    fn new() -> Queue<T> {
        Queue {
            front: None,
            back: std::ptr::null_mut(),
            size: 0,
        }
    }

    fn enqueue(&mut self, data: T) {
        let mut node = Box::new(Node { data, next: None });

        let raw = &mut *node as *mut Node<T>;

        if self.back.is_null() {
            self.front = Some(node);
            self.back = raw;
        } else {
            unsafe {
                (*self.back).next = Some(node);
            }

            self.back = raw;
        }

        self.size += 1;
    }

    fn dequeue(&mut self) -> Option<T> {
        let node = self.front.take()?;

        self.front = node.next;

        if self.front.is_none() {
            self.back = std::ptr::null_mut();
        }

        self.size -= 1;

        Some(node.data)
    }
}

fn main() {
    let mut q: Queue<i32> = Queue::new();
    q.enqueue(10); // FRONT -> [10] <- BACK
    q.enqueue(20); // FRONT -> [10] -> [20] <- BACK
    q.enqueue(30); // FRONT -> [10] -> [20] -> [30] <- BACK

    let a = q.dequeue(); // Some(10), FRONT -> [20] -> [30] <- BACK
    let b = q.dequeue(); // Some(20), FRONT -> [30] <- BACK
    let c = q.dequeue(); // Some(30), FRONT -> None <- BACK

    println!("Dequeued values: {:?}, {:?}, {:?}", a, b, c); // Output: Dequeued values: Some(10), Some(20), Some(30)
}
